### ISSUE TITLE
STAN vs NATS Timeout Defaults

### ISSUE THREAD

**Issue by samc1213** [nats-io/nats.c#404]
I understand STAN is built fully on top of NATS, and can, in some ways, be thought of as a separate entity. Thus, I understand that STAN has its own PING/PONG protocol, on top of NATS' PING/PONG logic.

However, as a new user of STAN, I found the default values rather confusing. In a controlled test, I simulated a "network partition" from the client to the NATS server, of a duration of about 20 seconds. I found that my STAN connection would "die" - I got a nice error callback from the subscription, however, I had to manually recreate the subscription (this behavior is fairly well documented [here ](https://docs.nats.io/nats-streaming-concepts/client-connections#note-important-things-to-know-about-reconnections))

The reason the STAN connection dies is because the STAN protocol, by default, tolerates 3 outstanding PINGs at an interval of 5 seconds:

https://github.com/nats-io/nats.c/blob/bfa8e54936fc1c59e3e1962b1d64e1223ad03d63/src/stan/copts.h#L24-L25

So, the math adds up that after 15 seconds, the STAN connection is dead.

However, by default, the underlying NATS connection does not reconnect until it notices 2 outstanding PINGs at an interval of two minutes.

https://github.com/nats-io/nats.c/blob/bfa8e54936fc1c59e3e1962b1d64e1223ad03d63/src/opts.h#L30-L31

Aren't these defaults a bit contradictory? This basically means that by default, STAN will give up before the underlying NATS connection tries again?

**Comment by kozlovic** [comment]
The comparison should not really be with NATS, but instead with the NATS Streaming server, which by default makes a streaming client connection survive around 7 minutes. I need to port the default STAN ping changes that I made for the Go client here: https://github.com/nats-io/stan.go/pull/337

**Comment by samc1213** [comment]
Thanks for the quick response @kozlovic . Just for my understanding, would you mind elaborating why we shouldn't compare to NATS, but rather to the Streaming server? I certainly understand comparing to the Streaming server, but why not also NATS? Seems like if STAN gives up before NATS, then any NATS issue would never recover

**Comment by kozlovic** [comment]
Take this setup:
```
STAN client -- (NATS connection to) --> NATS Server 1 <--- ROUTE ---> NATS Server 2 <--- STAN Server
```
Now suppose that the route breaks, STAN client's NATS connection will not be disconnected, the NATS pings will work just fine. Yet, STAN client has lost contact with the STAN Server. This is what STAN's Pings() are for and with default config will consider the STAN connection lost way too soon (this was a mistake on my part to use the default I used).
STAN Server will by default keep the STAN client connection "live" for over 7 minutes, yet the client will discard after less than 20 seconds.

As pointed out in the Go client PR, I have changed the default to be close to what the STAN server will use before considering a client lost.
